Gateway may panic when loading duplicate APIs on stable

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x42cfbb]
goroutine 405 [running]:
panic(0xccaae0, 0xc420010080)
/usr/local/go/src/runtime/panic.go:500 +0x1a1
main.processSpec(0xc42266c280, 0xc420411b30, 0x11, 0xc4252ae140, 0xc4252ae160, 0xc4252ae180, 0xc4253d0480, 0xc4253d04e0, 0xc420411b30, 0xc42003eee8)
/src/github.com/TykTechnologies/tyk/api_loader.go:149 +0x5bb
main.loadApps.func1(0xc4252a9a30, 0xc4200218d8, 0xc4252ae140, 0xc4252ae160, 0xc4252ae180, 0xc4253d0480, 0xc4253d04e0, 0xc4253d0540, 0xc42266c280, 0x11, ...)
/src/github.com/TykTechnologies/tyk/api_loader.go:646 +0x10a
created by main.loadApps
/src/github.com/TykTechnologies/tyk/api_loader.go:649 +0x4eb
```

That line contains:

```
prev := (*ApiSpecRegister)[referenceSpec.APIID]
```

Which was added in 2.3.6 as part of deduplicating APIs when loading them.

I'm fairly sure the panic refers to `ApiSpecRegister` being nil, and not `referenceSpec`, as the latter is used a few lines further up in the codebase. And `ApiSpecRegister` can be nil sometimes, as a pointer to a map.

The good news is that this is already fixed in master, as we got rid of the double pointer. We should do something similar in stable.

--- a/api.go
+++ b/api.go
@@ -58,24 +58,14 @@ func DoJSONWrite(w http.ResponseWriter, code int, responseMessage []byte) {
 func GetSpecForApi(APIID string) *APISpec {
 	apisMu.RLock()
 	defer apisMu.RUnlock()
-	if ApiSpecRegister == nil {
-		log.Error("No API Register present!")
-		return nil
-	}
-
-	spec, ok := (*ApiSpecRegister)[APIID]
-	if !ok {
-		return nil
-	}
-
-	return spec
+	return ApiSpecRegister[APIID]
 }
 
 func GetSpecForOrg(APIID string) *APISpec {
 	apisMu.RLock()
 	defer apisMu.RUnlock()
 	var aKey string
-	for k, v := range *ApiSpecRegister {
+	for k, v := range ApiSpecRegister {
 		if v.OrgID == APIID {
 			return v
 		}
@@ -83,7 +73,7 @@ func GetSpecForOrg(APIID string) *APISpec {
 	}
 
 	// If we can't find a spec, it doesn;t matter, because we default to Redis anyway, grab whatever you can find
-	return (*ApiSpecRegister)[aKey]
+	return ApiSpecRegister[aKey]
 }
 
 func checkAndApplyTrialPeriod(keyName string, apiId string, newSession *SessionState) {
@@ -151,7 +141,7 @@ func doAddOrUpdate(keyName string, newSession SessionState, dontReset bool) erro
 			log.Warning("No API Access Rights set, adding key to ALL.")
 			apisMu.RLock()
 			defer apisMu.RUnlock()
-			for _, spec := range *ApiSpecRegister {
+			for _, spec := range ApiSpecRegister {
 				if !dontReset {
 					spec.SessionManager.ResetQuota(keyName, newSession)
 					newSession.QuotaRenews = time.Now().Unix() + newSession.QuotaRenewalRate
@@ -448,7 +438,7 @@ func handleDeleteKey(keyName string, apiID string) ([]byte, int) {
 		// Go through ALL managed API's and delete the key
 		apisMu.RLock()
 		defer apisMu.RUnlock()
-		for _, spec := range *ApiSpecRegister {
+		for _, spec := range ApiSpecRegister {
 			spec.SessionManager.RemoveSession(keyName)
 			spec.SessionManager.ResetQuota(keyName, SessionState{})
 		}
@@ -515,7 +505,7 @@ func handleDeleteHashedKey(keyName string, apiID string) ([]byte, int) {
 		// Go through ALL managed API's and delete the key
 		apisMu.RLock()
 		defer apisMu.RUnlock()
-		for _, spec := range *ApiSpecRegister {
+		for _, spec := range ApiSpecRegister {
 			spec.SessionManager.RemoveSession(keyName)
 		}
 
@@ -616,10 +606,10 @@ func HandleGetAPIList() ([]byte, int) {
 
 	apisMu.RLock()
 	defer apisMu.RUnlock()
-	thisAPIIDList := make([]*tykcommon.APIDefinition, len(*ApiSpecRegister))
+	thisAPIIDList := make([]*tykcommon.APIDefinition, len(ApiSpecRegister))
 
 	c := 0
-	for _, apiSpec := range *ApiSpecRegister {
+	for _, apiSpec := range ApiSpecRegister {
 		thisAPIIDList[c] = apiSpec.APIDefinition
 		thisAPIIDList[c].RawData = nil
 		c++
@@ -641,7 +631,7 @@ func HandleGetAPI(APIID string) ([]byte, int) {
 
 	apisMu.RLock()
 	defer apisMu.RUnlock()
-	for _, apiSpec := range *ApiSpecRegister {
+	for _, apiSpec := range ApiSpecRegister {
 		if apiSpec.APIDefinition.APIID == APIID {
 
 			responseMessage, err = json.Marshal(apiSpec.APIDefinition)
@@ -1370,7 +1360,7 @@ func createKeyHandler(w http.ResponseWriter, r *http.Request) {
 
 					apisMu.RLock()
 					defer apisMu.RUnlock()
-					for _, spec := range *ApiSpecRegister {
+					for _, spec := range ApiSpecRegister {
 						checkAndApplyTrialPeriod(newKey, spec.APIID, &newSession)
 						if !spec.DontSetQuotasOnCreate {
 							// Reset quote by default

--- a/api_loader.go
+++ b/api_loader.go
@@ -146,7 +146,7 @@ func processSpec(referenceSpec *APISpec,
 		}
 		if !pathModified {
 			apisMu.RLock()
-			prev := (*ApiSpecRegister)[referenceSpec.APIID]
+			prev := ApiSpecRegister[referenceSpec.APIID]
 			apisMu.RUnlock()
 			if prev != nil && prev.Proxy.ListenPath == referenceSpec.Proxy.ListenPath {
 				// if this APIID was already loaded and
@@ -681,7 +681,7 @@ func loadApps(APISpecs *[]*APISpec, Muxer *mux.Router) {
 
 	// Swap in the new register
 	apisMu.Lock()
-	ApiSpecRegister = &tempSpecRegister
+	ApiSpecRegister = tempSpecRegister
 	apisMu.Unlock()
 
 	log.Debug("Checker host list")

--- a/api_test.go
+++ b/api_test.go
@@ -545,14 +545,14 @@ func TestGetOAuthClients(t *testing.T) {
 	var responseCode int
 
 	var tempSpecRegister = make(map[string]*APISpec)
-	ApiSpecRegister = &tempSpecRegister
+	ApiSpecRegister = tempSpecRegister
 
 	_, responseCode = getOauthClients(testAPIID)
 	if responseCode != 400 {
 		t.Fatal("Retrieving OAuth clients from nonexistent APIs must return error.")
 	}
 
-	(*ApiSpecRegister)[testAPIID] = &APISpec{}
+	ApiSpecRegister[testAPIID] = &APISpec{}
 
 	_, responseCode = getOauthClients(testAPIID)
 	if responseCode != 400 {

--- a/host_checker_manager.go
+++ b/host_checker_manager.go
@@ -215,7 +215,7 @@ func (hc *HostCheckerManager) OnHostDown(report HostHealthReport) {
 	hc.store.SetKey(hc.getHostKey(report), "1", int64(hc.checker.checkTimeout+1))
 
 	apisMu.RLock()
-	thisSpec, found := (*ApiSpecRegister)[report.MetaData[UnHealthyHostMetaDataAPIKey]]
+	thisSpec, found := ApiSpecRegister[report.MetaData[UnHealthyHostMetaDataAPIKey]]
 	apisMu.RUnlock()
 	if !found {
 		log.WithFields(logrus.Fields{
@@ -261,7 +261,7 @@ func (hc *HostCheckerManager) OnHostBackUp(report HostHealthReport) {
 	hc.store.DeleteKey(hc.getHostKey(report))
 
 	apisMu.RLock()
-	thisSpec, found := (*ApiSpecRegister)[report.MetaData[UnHealthyHostMetaDataAPIKey]]
+	thisSpec, found := ApiSpecRegister[report.MetaData[UnHealthyHostMetaDataAPIKey]]
 	apisMu.RUnlock()
 	if !found {
 		log.WithFields(logrus.Fields{
@@ -392,7 +392,7 @@ func (hc *HostCheckerManager) UpdateTrackingListByAPIID(hd []HostData, apiId str
 
 func (hc *HostCheckerManager) GetListFromService(APIID string) ([]HostData, error) {
 	apisMu.RLock()
-	spec, found := (*ApiSpecRegister)[APIID]
+	spec, found := ApiSpecRegister[APIID]
 	apisMu.RUnlock()
 	if !found {
 		return []HostData{}, errors.New("API ID not found in register!")
@@ -457,7 +457,7 @@ func (hc HostCheckerManager) RecordUptimeAnalytics(thisReport HostHealthReport) 
 	// If we are obfuscating API Keys, store the hashed representation (config check handled in hashing function)
 
 	apisMu.RLock()
-	thisSpec, found := (*ApiSpecRegister)[thisReport.MetaData[UnHealthyHostMetaDataAPIKey]]
+	thisSpec, found := ApiSpecRegister[thisReport.MetaData[UnHealthyHostMetaDataAPIKey]]
 	apisMu.RUnlock()
 	thisOrg := ""
 	if found {
@@ -527,7 +527,7 @@ func SetCheckerHostList() {
 	}).Info("Loading uptime tests...")
 	hostList := []HostData{}
 	apisMu.RLock()
-	for _, spec := range *ApiSpecRegister {
+	for _, spec := range ApiSpecRegister {
 		if spec.UptimeTests.Config.ServiceDiscovery.UseDiscoveryService {
 			thisHostList, sdErr := GlobalHostChecker.GetListFromService(spec.APIID)
 			if sdErr == nil {

--- a/main.go
+++ b/main.go
@@ -57,7 +57,7 @@ var argumentsBackup map[string]interface{}
 var DashService DashboardServiceSender
 
 var apisMu sync.RWMutex
-var ApiSpecRegister *map[string]*APISpec //make(map[string]*APISpec)
+var ApiSpecRegister map[string]*APISpec //make(map[string]*APISpec)
 var keyGen = DefaultKeyGenerator{}
 
 var mainRouter *mux.Router

